File: python/src/classifier.py
import numpy as np


class KNNClassifier:
    """
    k plus proches voisins, implementation maison (numpy pur).
    Pensee pour un port C++ ligne a ligne : pas de dependance ML externe.
    """

    def __init__(self, k=3, weighted=False, reject_threshold=None):
        """
        Args:
            k: nombre de voisins consideres
            weighted: si True, vote pondere par 1/distance (departage les egalites)
            reject_threshold: si la distance min depasse ce seuil, retourne None
                              (crop parasite / hors alphabet). None = pas de rejet.
        """
        self.k = k
        self.weighted = weighted
        self.reject_threshold = reject_threshold
        self.X_train = None   # (N, D) vecteurs de features
        self.y_train = None   # (N,)   labels

    def fit(self, X, y):
        """Memorise les donnees d'entrainement (k-NN n'a pas de vrai training)."""
        self.X_train = np.asarray(X, dtype=np.float32)
        self.y_train = np.asarray(y)

    def _squared_distances(self, x):
        """Distances au carre entre x (D,) et tous les points (sqrt inutile)."""
        return ((self.X_train - x) ** 2).sum(axis=1)

    def _k_nearest_indices(self, distances):
        """Indices des k plus proches points (pas forcement tries)."""
        k = min(self.k, len(distances))
        return np.argpartition(distances, k - 1)[:k]

    def _vote(self, neighbor_indices, distances):
        """
        Determine le label a partir des voisins.
        - vote simple : label majoritaire
        - vote pondere (self.weighted) : chaque voisin pese 1/distance
        """
        if self.weighted:
            weights = {}
            for idx in neighbor_indices:
                label = self.y_train[idx]
                weights[label] = weights.get(label, 0.0) + 1.0 / (distances[idx] + 1e-8)
            return max(weights, key=weights.get)
        labels = self.y_train[neighbor_indices]
        unique, counts = np.unique(labels, return_counts=True)
        return unique[np.argmax(counts)]

    def predict_one(self, x):
        """Predit le label d'un vecteur x (D,), ou None si rejet."""
        distances = self._squared_distances(x)
        if self.reject_threshold is not None and distances.min() > self.reject_threshold ** 2:
            return None
        idx = self._k_nearest_indices(distances)
        return self._vote(idx, distances)

    def predict(self, X):
        """Predit pour un vecteur unique (D,) ou un lot (M, D)."""
        X = np.asarray(X, dtype=np.float32)
        if X.ndim == 1:
            return self.predict_one(X)
        return np.array([self.predict_one(x) for x in X])

File: cpp/src/classifier.cc
#include "pyplus/classifier.hh"

#include <algorithm>
#include <map>
#include <numeric>
#include <stdexcept>
#include <utility>

namespace pyplus {

KNNClassifier::KNNClassifier(int k, bool weighted,
                             std::optional<float> reject_threshold)
    : k_(k), weighted_(weighted), reject_threshold_(reject_threshold) {}

void KNNClassifier::fit(Model model) { model_ = std::move(model); }

// Distances au carre entre x (d,) et tous les points (sqrt inutile).
std::vector<float> KNNClassifier::squared_distances(const float *x) const {
  std::vector<float> distances(static_cast<std::size_t>(model_.n));
  for (int i = 0; i < model_.n; ++i) {
    const float *row = model_.row(i);
    double acc = 0.0;
    for (int j = 0; j < model_.d; ++j) {
      const double diff = static_cast<double>(row[j]) - x[j];
      acc += diff * diff;
    }
    distances[static_cast<std::size_t>(i)] = static_cast<float>(acc);
  }
  return distances;
}

// Indices des k plus proches points (pas forcement tries).
std::vector<int>
KNNClassifier::k_nearest_indices(const std::vector<float> &distances) const {
  const int k = std::min<int>(k_, static_cast<int>(distances.size()));
  std::vector<int> indices(distances.size());
  std::iota(indices.begin(), indices.end(), 0);
  std::nth_element(indices.begin(), indices.begin() + (k - 1), indices.end(),
                   [&](int a, int b) {
                     return distances[static_cast<std::size_t>(a)] <
                            distances[static_cast<std::size_t>(b)];
                   });
  indices.resize(static_cast<std::size_t>(k));
  return indices;
}

// Determine le label a partir des voisins.
// - vote simple : label majoritaire (egalite -> premier par ordre trie,
//   comme np.unique + argmax)
// - vote pondere : chaque voisin pese 1/distance^2 (egalite -> premier
//   voisin rencontre, comme max() sur un dict Python)
std::string KNNClassifier::vote(const std::vector<int> &neighbor_indices,
                                const std::vector<float> &distances) const {
  if (weighted_) {
    std::vector<std::pair<std::string, double>> weights;
    for (int idx : neighbor_indices) {
      const std::string &label = model_.label(idx);
      auto it = std::find_if(weights.begin(), weights.end(),
                             [&](const auto &w) { return w.first == label; });
      if (it == weights.end())
        it = weights.insert(weights.end(), {label, 0.0});
      it->second +=
          1.0 / (static_cast<double>(distances[static_cast<std::size_t>(idx)]) +
                 1e-8);
    }
    const auto best = std::max_element(
        weights.begin(), weights.end(),
        [](const auto &a, const auto &b) { return a.second < b.second; });
    return best->first;
  }

  std::map<std::string, int> counts; // trie, comme np.unique
  for (int idx : neighbor_indices)
    ++counts[model_.label(idx)];
  const auto best = std::max_element(
      counts.begin(), counts.end(),
      [](const auto &a, const auto &b) { return a.second < b.second; });
  return best->first;
}

std::optional<std::string> KNNClassifier::predict_one(const float *x) const {
  if (model_.n == 0)
    throw std::runtime_error("classifieur non entraine (fit manquant)");

  const std::vector<float> distances = squared_distances(x);
  if (reject_threshold_) {
    const float min_dist =
        *std::min_element(distances.begin(), distances.end());
    if (min_dist > *reject_threshold_ * *reject_threshold_)
      return std::nullopt;
  }
  return vote(k_nearest_indices(distances), distances);
}

std::optional<std::string>
KNNClassifier::predict(const std::vector<float> &x) const {
  if (static_cast<int>(x.size()) != model_.d)
    throw std::runtime_error("dimension du vecteur != dimension du modele");
  return predict_one(x.data());
}

} // namespace pyplus
